electrum/txbatcher.py

```diff
@@ -0,0 +1,354 @@
+# This is denoted by tx2(tx1|o2).
+#
+# Example:
+#
+# output 1:     tx1(o1) ---------------
+#                                      \
+# output 2:     tx1'(o1,o2)-------      ----> tx2(tx1|o2) ------
+#                                 \     \                       \
+# output 3:     tx1''(o1,o2,o3)    \     ---> tx2'(tx1|o2,o3)    ---->  tx3(tx2|o3)  (if tx2 is mined)
+#                                   \
+#                                    -------------------------------->  tx3(tx1'|o3) (if tx1' is mined)
+#
+# In the above example, we have to make 3 payments.
+# Suppose we have broadcast tx1, tx1' and tx1''
+#  - if tx1 gets mined, we broadcast: tx2'(tx1|o2,o3)
+#  - if tx1' gets mined, we broadcast tx3(tx1'|o3)
+#
+# Note that there are two possible execution paths that may lead to the creation of tx3:
+#   - as a child of tx2
+#   - as a child of tx1'
+#
+# A batch is a set of incompatible txs, such as [tx1, tx1', tx1''].
+# Note that we do not persist older batches. We only persist the current batch in self.batch_txids.
+# Thus, if we need to broadcast tx2 or tx2', then self.batch_txids is reset, and the old batch is forgotten.
+#
+# If we cannot RBF a transaction (because the server returns an error), then we create a new batch,
+# as if the transaction had been mined.
+#   if cannot_rbf(tx1)  -> broadcast tx2(tx1,o2). The new base is now tx2(tx,o2)
+#   if cannot_rbf(tx1') -> broadcast tx3(tx1'|o3)
+#
+#
+# Notes:
+#
+# 1. CPFP:
+# When a batch is forgotten but not mined (because the server returned an error), we no longer bump its fee.
+# However, the current code does not theat the next batch as a CPFP when computing the fee.
+#
+# 2. Reorgs:
+# This code does not guarantee that a payment or a sweep will happen.
+# This is fine for sweeps; it is the responsibility of the caller (lnwatcher) to add them again.
+# To make payments reorg-safe, we would need to persist more data and redo failed payments.
+#
+# 3. batch_payments and batch_inputs are not persisted.
+# In the case of sweeps, lnwatcher ensures that SweepInfo is added again after a client restart.
+# In order to generalize that logic to payments, callers would need to pass a unique ID along with
+# the payment output, so that we can prevent paying twice.
+
+
+class TxBatcher(Logger):
+
+    SLEEP_INTERVAL = 1
+    RETRY_DELAY = 60
+
+    def __init__(self, wallet):
+        Logger.__init__(self)
+        self.wallet = wallet
+        self.config = wallet.config
+        self.db = wallet.db
+        self.lock = threading.Lock()
+        self.batch_payments = []       # list of payments we need to make
+        self.batch_inputs = {}         # list of inputs we need to sweep
+
+        # list of tx that were broadcast. Each tx is a RBF replacement of the previous one. Ony one can get mined.
+        self.batch_txids = self.db.get_stored_item("batch_txids", [])
+        self._base_tx = None           # current batch tx. last element of batch_txids
```

Comment:
TxBatcher needs to be taught the 100 KB max tx size standardness rule. It must not try creating txs that are too large.